# nf-core/eukavarizer: Citations

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools

- [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

> Andrews, S. (2010). FastQC: A Quality Control Tool for High Throughput Sequence Data [Online].

- [MultiQC](https://pubmed.ncbi.nlm.nih.gov/27312411/)

> Ewels P, Magnusson M, Lundin S, Käller M. MultiQC: summarize analysis results for multiple tools and samples in a single report. Bioinformatics. 2016 Oct 1;32(19):3047-8. doi: 10.1093/bioinformatics/btw354. Epub 2016 Jun 16. PubMed PMID: 27312411; PubMed Central PMCID: PMC5039924.

- [fastp](https://doi.org/10.1093/bioinformatics/bty560)

> Chen S, Zhou Y, Chen Y, Gu J. fastp: an ultra-fast all-in-one FASTQ preprocessor. Bioinformatics. 2018 Sep 1;34(17):i884-i890. doi: 10.1093/bioinformatics/bty560.

> Chen S. (2023). Ultrafast one-pass FASTQ data preprocessing, quality control, and deduplication using fastp. iMeta, 2:e107. https://doi.org/10.1002/imt2.107

- [fastplong]

> Shifu Chen. 2023. Ultrafast one-pass FASTQ data preprocessing, quality control, and deduplication using fastp. iMeta 2: e107. https://doi.org/10.1002/imt2.107

> Shifu Chen, Yanqing Zhou, Yaru Chen, Jia Gu; fastp: an ultra-fast all-in-one FASTQ preprocessor, Bioinformatics, Volume 34, Issue 17, 1 September 2018, Pages i884–i890, https://doi.org/10.1093/bioinformatics/bty560

- [SVYNC](https://github.com/nvnieuwk/svync)

> Vannieuwkerke N. (2024). SVYNC: Structural Variant Integrator. https://github.com/nvnieuwk/svync

- [BWA](https://doi.org/10.1093/bioinformatics/btp324)

> Li H, Durbin R. (2009). Fast and accurate short read alignment with Burrows-Wheeler transform. Bioinformatics, 25(14), 1754–1760.

- [BWA-MEM2](https://arxiv.org/abs/1907.12931)

> Vasimuddin M, Misra S, Li H, Aluru S. (2019). Efficient Architecture-Aware Acceleration of BWA-MEM for Multicore Systems. arXiv:1907.12931

- [Minimap2](https://doi.org/10.1093/bioinformatics/bty191)

> Li H. (2018). Minimap2: pairwise alignment for nucleotide sequences. Bioinformatics, 34(18), 3094-3100.

- [SAMtools](https://doi.org/10.1093/bioinformatics/btp352)

> Li H, Handsaker B, Wysoker A, et al. (2009). The Sequence Alignment/Map format and SAMtools. Bioinformatics, 25(16), 2078–2079.

- [BCFtools](https://doi.org/10.1093/gigascience/giab008)

> Danecek P, Bonfield JK, Liddle J, et al. (2021). Twelve years of SAMtools and BCFtools. GigaScience, 10(2), giab008.

- [BBMap](https://sourceforge.net/projects/bbmap/)

> Bushnell B. (2014). BBMap: A fast, accurate, splice-aware aligner. https://sourceforge.net/projects/bbmap/

- [seqtk](https://github.com/lh3/seqtk)

> Li H. (2012). seqtk: a fast and lightweight tool for processing sequences in the FASTA or FASTQ format. https://github.com/lh3/seqtk

- [DELLY](https://doi.org/10.1093/bioinformatics/bts378)

> Rausch T, Zichner T, Schlattl A, Stüzer A, Benes V, Korbel JO. (2012). DELLY: structural variant discovery by integrated paired-end and split-read analysis. Bioinformatics, 28(18), i333-i339.

- [GRIDSS](https://doi.org/10.1101/gr.222109.117)

> Cameron DL, Di Stefano L, Papenfuss AT. (2017). GRIDSS: sensitive and specific genomic rearrangement detection using positional de Bruijn graph assembly. Genome Res., 27(12), 2050-2060.

- [TIDDIT](https://doi.org/10.12688/f1000research.11501.1)

> Eisfeldt J, Vezzi F, Olason P, Nilsson D, Lindstrand A. (2017). TIDDIT, an efficient and comprehensive structural variant caller for massive parallel sequencing data. F1000Research, 6:664.

- [Dysgu](https://doi.org/10.1093/bioinformatics/btac144)

> Cawte N, Stadler PF, Simmonds MJ. (2022). Dysgu: a tool for structural variant detection from short-read sequencing data. Bioinformatics, 38(9), 2465-2471.

- [Sniffles](https://doi.org/10.1038/s41592-018-0001-7)

> Sedlazeck FJ, Rescheneder P, Smolka M, et al. (2018). Accurate detection of complex structural variations using single-molecule sequencing. Nat Methods, 15(6), 461-468.

- [CuteSV](https://doi.org/10.1186/s13059-020-02107-y)

> Jiang T, Liu Y, Jiang Y, et al. (2020). Long-read-based human genomic structural variation detection with cuteSV. Genome Biol., 21:189.

- [SvABA](https://doi.org/10.1101/gr.221028.117)

> Wala JA, Bandopadhayay P, Greenwald NF, et al. (2018). SvABA: genome-wide detection of structural variants and indels by local assembly. Genome Res., 28(4), 581–591.

- [Manta](https://doi.org/10.1093/bioinformatics/btv710)

> Chen X, Schulz-Trieglaff O, Shaw R, et al. (2016). Manta: rapid detection of structural variants and indels for germline and cancer sequencing applications. Bioinformatics, 32(8), 1220–1222.

- [SURVIVOR](https://doi.org/10.1038/ncomms14061)

> Jeffares DC, Jolly C, Hoti M, et al. (2017). Transient structural variations have strong effects on quantitative traits and reproductive isolation in fission yeast. Nat Commun., 8, 14061.

- [Varify]

> Sloup O. et al. (2024). Varify: structural variant analysis summary and visualization. https://github.com/Lupphes/Varify

- [SRA](https://doi.org/10.1093/nar/gkq1019)

> Leinonen R, Sugawara H, Shumway M; International Nucleotide Sequence Database Collaboration. (2011). The Sequence Read Archive. Nucleic Acids Res., 39(Database issue): D19–D21.

- [BioDbCore](https://github.com/luppo/biodbcore)

> Sloup O. (2024). BioDbCore: ENA/RefSeq retrieval module. https://github.com/luppo/biodbcore

- [coreutils](https://www.gnu.org/software/coreutils/)

> GNU Project. coreutils (including gunzip). https://www.gnu.org/software/coreutils/
